postgres/datadog_checks/postgres/statements.py

```diff
@@ -0,0 +1,181 @@
+        version is not a reliable way to determine the available columns on `pg_stat_statements`. The database can
+        be upgraded without upgrading extensions, even when the extension is included by default.
+        """
+        query = """
+            SELECT column_name
+            FROM information_schema.columns
+            WHERE table_schema = 'public'
+            AND table_name = 'pg_stat_statements';
+            """
+        columns = self._execute_query(db.cursor(), query)
+        return [column[0] for column in columns]
+
+    def collect_per_statement_metrics(self, instance, db, instance_tags):
+        try:
+            self.__collect_per_statement_metrics(instance, db, instance_tags)
+        except Exception:
+            db.rollback()
+            logger.exception('Unable to collect statement metrics due to an error')
+
+    def __collect_per_statement_metrics(self, instance, db, instance_tags):
+        available_columns = self._get_pg_stat_statements_columns(db)
+        missing_columns = PG_STAT_STATEMENTS_REQUIRED_COLUMNS - set(available_columns)
+        if len(missing_columns) > 0:
+            logger.warning(
+                'Unable to collect statement metrics because required fields are unavailable: %s',
+                ', '.join(list(missing_columns)),
+            )
+            return
+
+        desired_columns = (
+            list(PG_STAT_STATEMENTS_METRIC_COLUMNS.keys())
+            + list(PG_STAT_STATEMENTS_OPTIONAL_COLUMNS)
+            + list(PG_STAT_STATEMENTS_TAG_COLUMNS.keys())
+        )
+        query_columns = list(
+            set(desired_columns) & set(available_columns) + list(PG_STAT_STATEMENTS_TAG_COLUMNS.keys())
+        )
+        rows = self._execute_query(
+            db.cursor(cursor_factory=psycopg2.extras.DictCursor),
+            STATEMENTS_QUERY.format(
+                cols=', '.join(query_columns),
+                pg_stat_statements_view=self.config.pg_stat_statements_view,
+            ),
+            params=(self.config.dbname,),
+        )
+        if not rows:
+            return
+
+        def row_keyfunc(row):
+            # old versions of pg_stat_statements don't have a query ID so fall back to the query string itself
+            queryid = row['queryid'] if 'queryid' in row else row['query']
+            return (queryid, row['datname'], row['rolname'])
+
+        rows = self._state.compute_derivative_rows(rows, PG_STAT_STATEMENTS_METRIC_COLUMNS.keys(), key=row_keyfunc)
+        metric_limits = (
+            self.config.statement_metric_limits if self.config.statement_metric_limits else DEFAULT_METRIC_LIMITS
+        )
+        rows = apply_row_limits(rows, metric_limits, 'calls', True, key=row_keyfunc)
+
+        for row in rows:
+            try:
+                normalized_query = datadog_agent.obfuscate_sql(row['query'])
+                if not normalized_query:
+                    logger.warning("Query obfuscation resulted in empty query '%s'", row['query'])
```

Comment:
In this case for postgres `pg_stat_statements`, yes it is. IMO this is an issue upstream in the agent core and how we treat obfuscation/normalization as the same thing. Here we are using a function called obfuscate_sql, but what we really care about is normalization (we don't actually care about obfuscation because parameters are already obfuscated when stats are computed on the table).

When we collect query samples, however, we will care about obfuscation failing and we will reject that query sample. I'll add a comment explaining this.